### PR TITLE
piwww: testrun - add proposal routes.

### PR DESCRIPTION
**Pull request by amass01**

### CODE REVIEW

File: politeiawww/cmd/piwww/testrun.go [hunk 1]
```diff
@@ -446,15 +470,217 @@ func testUserRoutes(admin testUser, minPasswordLength int) error {
...
+
+	return &pi.ProposalNew{
+		Files:     files,
+		Metadata:  metadata,
+		PublicKey: hex.EncodeToString(cfg.Identity.Public.Key[:]),
+		Signature: sig,
+	}, nil
+}
+
+// testPropsWithComments tests the propsal & comments routes
```
Comment (amass01) [inline politeiawww/cmd/piwww/testrun.go]:
```suggestion
// testProposalRoutes tests the propsal routes
```

Will do in comment routes PR.

File: politeiawww/cmd/piwww/testrun.go [hunk 2]
```diff
@@ -446,15 +470,217 @@ func testUserRoutes(admin testUser, minPasswordLength int) error {
...
+
+	// Create test user
+	fmt.Printf("Creating test user\n")
+	user, err := createUserAndVerifyEmail(minPasswordLength, false)
+	if err != nil {
+		return err
+	}
+
+	// Submit new proposal
+	fmt.Printf("  New proposal\n")
```
Comment (amass01) [inline politeiawww/cmd/piwww/testrun.go:539]: This was wrapped in a standalone func on #50, as we submit a new proposal and then use it to run comment routes.

**Review by lukebp — CHANGES_REQUESTED** [review]
These need to be more comprehensive.  Make sure to test the common unvetted and vetted paths.

Submit new proposal
Edit unvetted proposal
Censor unvetted proposal
Submit new proposal
Make proposal public
Edit vetted proposal
Censor vetted proposal
Submit new proposal
Make proposal public
Abandon vetted proposal
Submit new proposal and leave it unvetted
Get proposal inventory
Get unvetted proposals
Get vetted proposals

File: politeiawww/cmd/piwww/testrun.go [hunk 3]
```diff
@@ -24,13 +29,20 @@ type testRunCmd struct {
 	} `positional-args:"true" required:"true"`
 }
 
+var (
+	minPasswordLength int
```
Comment (lukebp) [inline politeiawww/cmd/piwww/testrun.go:32]: Why do you need to use globals here?

File: politeiawww/cmd/piwww/testrun.go [hunk 4]
```diff
@@ -108,54 +120,39 @@ func userManage(userID, action, reason string) error {
...
-		// Test users
-		user testUser
-	)
-	// Run user routes.
-	fmt.Printf("Running user routes\n")
-
+// createUserAndVerifyEmail creates new user, tries to resend email verification
+// if resendEmail flag is true, verfies user's email, logs in as the newly
+// created user and finally returns the created testUser
+func createUserAndVerifyEmail(minPasswordLength int, resendEmail bool) (*testUser, error) {
```
Comment (lukebp) [inline politeiawww/cmd/piwww/testrun.go]: This function is doing way too many things.  A function should only do one thing.  You're using it to do multiple things: create a new user, test the verification resend route, and test the update user key route.  Do not do this.  If you create a function to create a new user then it should not be testing additional routes in it.

File: politeiawww/cmd/piwww/testrun.go [hunk 5]
```diff
@@ -446,15 +475,436 @@ func testUserRoutes(admin testUser, minPasswordLength int) error {
...
+		Files:     files,
+		Metadata:  metadata,
+		PublicKey: hex.EncodeToString(cfg.Identity.Public.Key[:]),
+		Signature: sig,
+	}, nil
+}
+
+// submitNewPropsal submits new proposal and verifies it if given verify bool
+// is true
+func submitNewProposal(verify bool) (string, error) {
```
Comment (lukebp) [inline politeiawww/cmd/piwww/testrun.go]: You don't need a bool for `verify`.  The client should always be verifying the returned response.

File: politeiawww/cmd/piwww/testrun.go [hunk 6]
```diff
@@ -446,15 +475,436 @@ func testUserRoutes(admin testUser, minPasswordLength int) error {
...
+	const proposalStatusCensored = "censored"
+	err := proposalSetStatus(token, proposalStatusCensored, reason, unvetted)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// proposalPublic makes given proposal public
+func proposalPublic(token string) error {
```
Comment (lukebp) [inline politeiawww/cmd/piwww/testrun.go]: This is too specific of a function.  You should have a `proposalSetStatus` function that allows you to set the status to whatever you want so you don't need a new function for every status.

File: politeiawww/cmd/piwww/testrun.go [hunk 7]
```diff
@@ -446,15 +475,436 @@ func testUserRoutes(admin testUser, minPasswordLength int) error {
...
+	const proposalStatusAbandoned = "abandoned"
+	err := proposalSetStatus(token, proposalStatusAbandoned, reason, false)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// proposalEdit edits given proposal
+func proposalEdit(token string, unvetted, random bool) error {
```
Comment (lukebp) [inline politeiawww/cmd/piwww/testrun.go]: What's the purpose of `random`?  If its set to false this function is going to error.  It has to be true, so why do you need it as an argument?

Comment (lukebp) [inline politeiawww/cmd/piwww/testrun.go]: I would prefer this to be:
```suggestion
func proposalEdit(state pi.PropStateT, token string) error {
```

For two reasons:
1. Listing the state first is how the data model is setup.  Specify the state (which corresponds to a db on the backend) then specify the token (which corresponds to a db entry in that db).
2. Using `pi.PropStateT` makes the function call more descriptive: `proposalEdit(pi.PropStateUnvetted, token)` vs `proposalsEdit(true, token)`

File: politeiawww/cmd/piwww/testrun.go [hunk 8]
```diff
@@ -446,15 +475,436 @@ func testUserRoutes(admin testUser, minPasswordLength int) error {
...
+	epc.Args.Token = token
+	err := epc.Execute(nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// proposals fetchs requested proposals and verifies returned map length
+func proposals(ps pi.Proposals) error {
```
Comment (lukebp) [inline politeiawww/cmd/piwww/testrun.go]: Why do you have a function that fetches proposals, but that doesn't return the proposals that it fetched?  That doesn't make any sense.

**Review by lukebp — CHANGES_REQUESTED** [review]
This PR needs some work.  Go through the cms testrun code and look at how things are setup.  That's what I want for this code too.  Also, a general comment is that your functions are too specific.  Functions should do only 1 thing and should be generic functions that map to the data model.  See my comments for more details.

File: politeiawww/cmd/piwww/testrun.go (hunk 3 above)
Comment (amass01) [reply politeiawww/cmd/piwww/testrun.go:32]: I thought it's better than passing them around also to nested functions.

File: politeiawww/cmd/piwww/testrun.go (hunk 7 above)
Comment (amass01) [reply politeiawww/cmd/piwww/testrun.go]: Totally agree

File: politeiawww/cmd/piwww/testrun.go (hunk 6 above)
Comment (amass01) [reply politeiawww/cmd/piwww/testrun.go]: It's just a wrapper, i have already `proposalSetStatus` but also added proposalPublic, proposalCensor & proposalAbandon as wrappers.

File: politeiawww/cmd/piwww/testrun.go [hunk 9]
```diff
@@ -446,15 +476,434 @@ func testUserRoutes(admin testUser, minPasswordLength int) error {
...
+	if err != nil {
+		return "", err
+	}
+
+	return token, nil
+}
+
+// proposalSetStatus calls proposal set status command
+//
+// This function returns with user logged out
```
Comment (amass01) [inline politeiawww/cmd/piwww/testrun.go:576]:
```suggestion
// This function returns with the user logged out
```

Comment (amass01) [reply politeiawww/cmd/piwww/testrun.go:576]: updated this in #50

**Review by lukebp — CHANGES_REQUESTED** [review]
The testrun command should fetch data using both the full sized token and the short token.  This should be applied to proposals, comments, and votes.  Anytime it fetches data from the pi api.

**Review by lukebp — APPROVED** [review]
ACK

**Pull request merged**